crates/mun_memory/src/diff.rs

```diff
@@ -174,117 +190,133 @@ fn append_struct_mapping(
+                            deletion_idx: *deletion_idx,
+                            insertion_idx: *insertion_idx,
+                            old_index: *old_idx,
+                            new_index: *new_idx,
+                            old_ty: old_ty.clone(),
+                            new_ty: new_ty.clone(),
+                            old_fields,
+                            new_fields,
+                            length,
+                        })
                     } else {
-                        // `std::usize::MAX` is used to indicate that the respective two fields are
-                        // too different.
-                        std::usize::MAX
+                        // Indicate that the respective two fields are too different.
+                        None
                     }
                 })
-                .collect::<Vec<usize>>()
+                .collect::<Vec<LengthDescription>>()
         })
         .collect();
 
-    let mut used_deletions = vec![false; num_deleted];
-    let mut used_insertions = vec![false; num_inserted];
-    // Traverse all (insertion, deletion) pairs in ascending order of their `myers::diff_length`.
-    while let Some((idx, length)) = myers_lengths.iter().enumerate().min_by(|x, y| x.1.cmp(y.1)) {
-        // Skip marked fields
-        if *length == std::usize::MAX {
-            break;
-        }
-
-        let delete_idx = idx % num_deleted;
-        unsafe { *used_deletions.get_unchecked_mut(delete_idx) = true };
-
-        let insert_idx = idx / num_deleted;
-        unsafe { *used_insertions.get_unchecked_mut(insert_idx) = true };
-
-        let old_index = unsafe { *deletions.get_unchecked(delete_idx) };
-        let new_index = unsafe { *insertions.get_unchecked(insert_idx) };
-        // If there is no difference between the old and new fields
-        mapping.push(if *length == 0 {
-            // Move the struct
-            Diff::Move {
-                old_index,
-                new_index,
-            }
-        } else {
-            let old_fields = unsafe { old_fields.get_unchecked(old_index) };
-            let new_fields = unsafe { new_fields.get_unchecked(new_index) };
-
-            // ASSUMPTION: Don't use recursion, because all types are individually checked for
-            // differences.
-            // TODO: Support value struct vs heap struct?
-            let diff = field_diff(old_fields, new_fields);
-
-            // Edit the struct, potentially moving it in the process.
-            Diff::Edit {
-                diff,
-                old_index,
-                new_index,
+    // Sort in ascending order of their `myers::diff_length`.
+    myers_lengths.sort_by(
+        |LengthDescription { length: lhs, .. }, LengthDescription { length: rhs, .. }| lhs.cmp(rhs),
+    );
+
+    let mut used_deletions = vec![false; deletions.len()];
+    let mut used_insertions = vec![false; insertions.len()];
+    myers_lengths.into_iter().for_each(
```

Comment:
This is just a regular `for` statement. I would write it as such to reduce the indentation needed.